parachain/pallets/erc20-app/src/payload.rs

```diff
@@ -28,10 +30,28 @@ pub struct InPayload<AccountId: codec::Decode> {
 
 impl<AccountId: codec::Decode> InPayload<AccountId> {
 
-	/// Decode payload into an Ethereum log event and then extract ABI-encoded data from log event
-	pub fn decode(payload: &[u8]) -> Result<Self, DecodeError> {
+	/// Decode payload into an Ethereum log event, check that it matches the verified receipt,
+	/// and then extract ABI-encoded data from log event
+	pub fn decode_verified(payload: &[u8], verification_output: &VerificationOutput) -> Result<Self, DecodeError> {
 		// Decode ethereum Log event from RLP-encoded data
 		let log: Log = rlp::decode(payload)?;
+
+		if let VerificationOutput::Receipt(receipt) = verification_output {
+			if !receipt.contains_log(&log) {
+				return Err(DecodeError::InvalidVerification);
```

Comment:
Strictly speaking a failed verification wouldn't be a decoding error. Maybe `InvalidVerification` would be better defined as a pallet error (which ties in somewhat with my previous point about moving this code to `Application::handle`).

Reply:
Agreed. I put this in `InPayload` to reduce the likelihood of someone inadvertently dropping the verification check when they change the code. But hiding it in `InPayload` isn't a pretty poor strategy